node/core/approval-voting/src/lib.rs

```diff
@@ -683,9 +690,21 @@ where
 
 	let mut last_finalized_height: Option<BlockNumber> = None;
 
+	let startup_delay =
+		|| async move { future::pending::<()>().timeout(STARTUP_WAIT_DURATION).await };
+
 	loop {
 		let mut overlayed_db = OverlayedBackend::new(&backend);
 		let actions = futures::select! {
+			_res = startup_delay().fuse() => {
+				if !state.recovery_state.complete() {
```

Comment:
The whole `RecoveryState` thing seems like a workaround to make sure that we only actually do this `handle_startup` work on the first iteration of the loop.

I see that you create the timer anew on each iteration of the loop, but I think there has to be some solution where you create a future which only gives a result the first time. 2 possible solutions.

1. Instantiate `startup_delay` as a simple `futures-timer` `Delay`.
2. Inside the loop, invoke `(&mut startup_delay).fuse()`

The other solution is to declare `startup_delay` as a `stream::once(Delay).chain(stream::empty)`

Reply:
Amazing... I was trying to craft a manual future that did exactly what stream::once(Delay).chain(stream::empty) does.